core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java

```diff
@@ -35,40 +35,55 @@ public String getName()
     public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceOutput, Block block)
     {
         AbstractRowBlock rowBlock = (AbstractRowBlock) block;
+        boolean isNullPresent = rowBlock.mayHaveNull();
+        int[] fieldBlockOffsets = rowBlock.getFieldBlockOffsets();
+
+        if (isNullPresent && fieldBlockOffsets == null) {
+            throw new IllegalStateException("fieldBlockOffsets cannot be null when null value is present");
+        }
+
         int numFields = rowBlock.numFields;
 
         int positionCount = rowBlock.getPositionCount();
 
         int offsetBase = rowBlock.getOffsetBase();
-        int[] fieldBlockOffsets = rowBlock.getFieldBlockOffsets();
-        int startFieldBlockOffset = fieldBlockOffsets[offsetBase];
-        int endFieldBlockOffset = fieldBlockOffsets[offsetBase + positionCount];
+
+        int startFieldBlockOffset = fieldBlockOffsets != null ? fieldBlockOffsets[offsetBase] : offsetBase;
+        int endFieldBlockOffset = fieldBlockOffsets != null ? fieldBlockOffsets[offsetBase + positionCount] : offsetBase + positionCount;
 
         sliceOutput.appendInt(numFields);
+        sliceOutput.appendInt(positionCount);
+
         for (int i = 0; i < numFields; i++) {
             blockEncodingSerde.writeBlock(sliceOutput, rowBlock.getRawFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
         }
 
-        sliceOutput.appendInt(positionCount);
-        for (int position = 0; position < positionCount + 1; position++) {
-            sliceOutput.writeInt(fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset);
-        }
         EncoderUtil.encodeNullsAsBits(sliceOutput, block);
+
+        if (isNullPresent) {
+            for (int position = 0; position < positionCount + 1; position++) {
```

Comment:
nit: most likely it makes sense to compute `fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset` for all positions, then use `writeBytes(wrappedIntArray(...))` instead of appending `writeInt` positions by positions

See how other serializers work (e.g. `io.trino.spi.block.IntArrayBlockEncoding`